Реализация первой задачи

Jejevkin opened the pull request:
Описание проекта в README.md

Commit:
Redis added

Commit:
Implemented endpoints for places search and favorites management

Commit:
Add functional tests; Code refactoring

Commit:
Updated README.md

Ivan1586 commented on src/api/v1/places.py:
```diff
@@ -0,0 +1,125 @@
+
+from core.common import request_key_builder
+from core.config import settings
+from schemas.places import (NearbyPlaceRequest, NearbyPlaceResponse, SearchPlaceRequest,
+                            SearchPlaceResponse, FavoritePlaceCreate, FavoritePlaceResponse)
```
Лучше выстраивать импорты таким образом, так более читаемо
```
from schemas.places import (
    NearbyPlaceRequest,
    NearbyPlaceResponse,
    SearchPlaceRequest,
    SearchPlaceResponse,
    эFavoritePlaceCreate,
    FavoritePlaceResponse
)
```

Jejevkin replied:
исправил

Ivan1586 commented on tests/ft/src/test_auth.py:
```diff
@@ -0,0 +1,58 @@
+    ]
+)
+@pytest.mark.asyncio
+async def test_signup(make_post_request, user_data, expected_status, expected_detail):
```
Давай все тесты чутка перепишем по паттерну ААА - будет хорошей практикой

Jejevkin replied:
Исправил

Ivan1586 commented on tests/ft/run_tests.sh:
```diff
@@ -0,0 +1,4 @@
+set -ex
+docker-compose up -d
+sleep 10
+pytest . 1> log.log 2>error.log
```
No newline at the end of file

Jejevkin replied:
исправил

Ivan1586 commented on tests/ft/.env.sample:
```diff
@@ -0,0 +1,24 @@
+POSTGRES_DB=travel_db
+POSTGRES_USER=app
+POSTGRES_PASSWORD=123qwe
+
+PROJECT_NAME='Travel Companion'
+API_VERSION=v1
+AUTHJWT_SECRET_KEY=secret
+AUTHJWT_ALGORITHM=HS256
+
+PSQL_HOST='postgres'
+PSQL_PORT=5432
+PSQL_USER='app'
+PSQL_PASSWORD='123qwe'
+PSQL_DB='travel_db'
+DB_ENGINE_ECHO=False
+
+REDIS_HOST=redis
+REDIS_PORT=6379
+REDIS_TTL=300
+
+LOCATIONIQ_API_KEY=key
+LOCATIONIQ_BASE_URL='https://eu1.locationiq.com/v1'
```
Обычно при заливе в гит эти данные не заполняют:
```
POSTGRES_DB=
POSTGRES_USER=
POSTGRES_PASSWORD=

PROJECT_NAME=
API_VERSION=
AUTHJWT_SECRET_KEY=
AUTHJWT_ALGORITHM=

PSQL_HOST=
PSQL_PORT=
PSQL_USER=
PSQL_PASSWORD=
PSQL_DB=
DB_ENGINE_ECHO=

REDIS_HOST=
REDIS_PORT=
REDIS_TTL=

LOCATIONIQ_API_KEY=
LOCATIONIQ_BASE_URL=
```

Jejevkin replied:
исправил

Ivan1586 commented on src/services/place.py:
```diff
@@ -0,0 +1,145 @@
+from db.database import get_session
+from db.redis import get_redis
+from models.places import Place, SearchHistory, FavoritePlace
+from schemas.places import NearbyPlaceRequest, NearbyPlaceResponse, SearchPlaceRequest, SearchPlaceResponse, \
```
Сгруппируй импорты более человекочитаемо, как я уже писал

Jejevkin replied:
исправил

Ivan1586 commented on src/services/place.py:
```diff
@@ -0,0 +1,145 @@
+    def __init__(self, db: AsyncSession, redis: Redis):
+        super().__init__(db)
+        self.redis = redis
+        self.client = httpx.AsyncClient()
```
Тут могу ошибаться, но при завершении работы приложения не нужно закрывать клиент (например, через вызов `client.aclose()`), чтобы освободить ресурсы?

Jejevkin replied:
Все верно. Дока неоднократно предупреждает об этом. Буду исправлять

Jejevkin replied:
исправил

Ivan1586 commented on src/services/place.py:
```diff
@@ -0,0 +1,145 @@
+
+        if favorite_place:
+            return await self._delete_entity(favorite_place)
+        logger.warning(f'Не удалось удалить место {place_id} из избранного пользователя {user_id}.')
```
Стоит передать более подробную информацию в ответ, чтобы клиент тоже знал, почему `delete` ничего не удалил

Jejevkin replied:
Я выводил эту информацию на уровне api:
`raise HTTPException(HTTPStatus.NOT_FOUND, 'Favorite place not found')`

Ivan1586 commented on src/services/place.py:
```diff
@@ -0,0 +1,145 @@
+        Выполняет запрос к API LocationIQ и возвращает данные в формате JSON.
+        """
+        response = await self.client.get(url, params=params)
+        response.raise_for_status()
```
Есть вызов `raise_for_status`, но стоит явно обрабатывать такие исключения, чтобы при возникновении HTTP ошибок вернуть пользователю понятный ответ

Jejevkin replied:
исправил

Ivan1586 commented on src/services/place.py:
```diff
@@ -0,0 +1,145 @@
+                entities_to_save.append(history_record)
+
+        if saved_entities := await self._save_entities(entities_to_save):
+            logger.info(f'Данные {saved_entities} успешно сохранены в базе данных.')
```
Будто такие логи могут стать слишком подробными, если сохранено большое количество объектов. Лучше просто указать общее количество сохраненных записей

Jejevkin replied:
исправил

Ivan1586 commented on src/services/place.py:
```diff
@@ -0,0 +1,145 @@
+                    user_id=user_id,
+                    place_id=new_place.place_id,
+                )
+                entities_to_save.append(history_record)
```
Есть риск race condition, если `Place` уже существует в базе данных?

Jejevkin replied:
У меня стоит уникальный индекс на place_id. В случае race condition Place не сохранится, но это не страшно, так как запись уже будет в базе. А вот пользователь рискует остаться без истории поиска, потому что вся транзакция откатится. Буду исправлять

Jejevkin replied:
исправил

Ivan1586 commented on src/models/places.py:
```diff
@@ -0,0 +1,67 @@
+class Place(Base):
+    __tablename__ = 'places'
+
+    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4, unique=True, nullable=False)
```
Не требуется дополнительно указывать `unique=True` для первичных ключей, так как они уже уникальны. Можно убрать эту опцию, чтобы сократить количество избыточных проверок на уровне базы данных

Jejevkin replied:
исправил

Ivan1586 commented on src/models/places.py:
```diff
@@ -0,0 +1,67 @@
+    place_id = Column(String, unique=True, nullable=False)
+    lat = Column(Float, nullable=False)
+    lon = Column(Float, nullable=False)
+    display_name = Column(String)
```
Насколько длинными могут быть значения здесь (например, 255 символов?). Если такие данные предполагаются длинными, можно рассмотреть использование `Text` вместо `String`

Jejevkin replied:
исправил

Ivan1586 commented on src/models/places.py:
```diff
@@ -0,0 +1,67 @@
+    created_at = Column(DateTime, default=datetime.utcnow)
+
+    @classmethod
+    def from_schema(cls, place: 'BasePlaceResponse') -> 'Place':
+        return cls(
+            place_id=place.place_id,
+            lat=place.lat,
+            lon=place.lon,
+            display_name=place.display_name,
+            place_class=place.place_class,
+            place_type=place.place_type,
+        )
+
+    def __repr__(self) -> str:
+        return f'<Place {self.id}>'
```
Как будто нужна обработка значений (например, проверка на их корректность), лучше добавить валидацию или обработку исключений. Например:
```
  try:
          place_id = place.place_id.strip()
          ...
      except AttributeError:
          raise ValueError("Некорректные данные для создания Place.")
```

Jejevkin replied:
Валидация проходит до сопоставления с моделью и записи в базу:
` validated_items = [model.model_validate(item) for item in raw_data]`

Добавил дополнительные ограничения на уровне pydantic моделей

Ivan1586 commented on src/models/places.py:
```diff
@@ -0,0 +1,67 @@
+    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4, unique=True, nullable=False)
+    user_id = Column(UUID(as_uuid=True), ForeignKey('users.id'), nullable=False)
+    place_id = Column(String, ForeignKey('places.place_id'), nullable=False)
+    search_date = Column(DateTime, default=datetime.utcnow)
```
**nit:** Имеет смысл добавить `onupdate=datetime.utcnow`, если есть планы обновлять эту запись

Jejevkin replied:
Добавил. Скорее всего в дальнейшем пригодится для выполнения 3го задания

Ivan1586 commented on src/models/places.py:
```diff
@@ -0,0 +1,67 @@
+from schemas.places import BasePlaceResponse
+
+
+class Place(Base):
```
На уровне модели отсутствуют дополнительные проверки, например, на допустимый диапазон широты и долготы (lat, lon)

Jejevkin replied:
исправил

Commit:
Fix review suggestions:
- Regroup imports
- Add missing newline at EOF
- Remove .env.sample variables
- Add explicit client.aclose() call

Commit:
Fix review suggestions:
- Add external service exception handler

Commit:
Fix review suggestions:
- Improve validations

Commit:
Fix review suggestions:
- Change tests for AAA pattern
